### PR TITLE
Cross-dataset: Corrected EMA warmup (timm-style) — fix decay=0.999 at step 750 bug (CODE CHANGE)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current EMA implementation uses a fixed `decay=0.9999` from step 0. At step 750 (early training), the effective weight on the EMA model is 0.9999^750 ≈ 0.928, meaning the EMA is already 93% of current weights — but the model hasn't converged yet, so the EMA is accumulating early-training noise from the very first steps.

The timm-style EMA warmup fix uses an adaptive decay schedule:

```
actual_decay = min(target_decay, (1 + step) / (10 + step))
```

At step 1: actual_decay = 2/11 ≈ 0.18 (EMA tracks current model closely)
At step 100: actual_decay = 101/110 ≈ 0.918
At step 1000: actual_decay = 1001/1010 ≈ 0.991
At step 10000+: actual_decay ≈ target_decay = 0.9999

This prevents the EMA from being "poisoned" by early random-ish weights and lets it converge to a proper ensemble of late-training checkpoints. The current codebase uses `--no-use-ema` in all baselines precisely because EMA was broken; this fix should make EMA useful again.

## Instructions

**CODE CHANGE REQUIRED** — replace the EMA class in `train.py` with the warmup-corrected version.

### Step 1: Replace the EMA class

Find the existing EMA class (search for `class EMA` or `ema_decay`) and replace with:

```python
import copy

class EMAWithWarmup:
    def __init__(self, model, decay=0.9999):
        self.model = copy.deepcopy(model)
        self.decay = decay
        self.step = 0

    def update(self, model):
        self.step += 1
        actual_decay = min(self.decay, (1 + self.step) / (10 + self.step))
        with torch.no_grad():
            for ema_p, p in zip(self.model.parameters(), model.parameters()):
                ema_p.mul_(actual_decay).add_(p, alpha=1 - actual_decay)

    def state_dict(self):
        return self.model.state_dict()
```

### Step 2: Remove `--no-use-ema` from runs — use EMA for all experiments

Use `--wandb_group ema-warmup` for all runs.

**DrivAerML — EMA warmup decay=0.9999:**
```bash
python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --wandb_group ema-warmup
```

**DrivAerML — EMA warmup decay=0.999:**
```bash
python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --ema-decay 0.999 --wandb_group ema-warmup
```

**AirfRANS — EMA warmup decay=0.9999:**
```bash
python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --wandb_group ema-warmup
```

**AirfRANS — EMA warmup decay=0.999:**
```bash
python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --ema-decay 0.999 --wandb_group ema-warmup
```

**TandemFoil — EMA warmup decay=0.9999:**
```bash
python train.py --dataset tandemfoil --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --wandb_group ema-warmup
```

**TandemFoil — EMA warmup decay=0.999:**
```bash
python train.py --dataset tandemfoil --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --ema-decay 0.999 --wandb_group ema-warmup
```

Note: if the `--ema-decay` flag doesn't exist yet, add it:
```python
parser.add_argument('--ema-decay', type=float, default=0.9999)
```

Run 6 experiments total (2 per dataset: decay 0.9999 and 0.999). Use the remaining 2 GPU slots to run the two most promising datasets again if one clearly outperforms.

## Baseline

Current best metrics (all using `--no-use-ema`):
- **TandemFoil**: val/loss = 30.10 (epoch 157, PR #2810, W&B run v6amjkh7)
- **AirfRANS**: val/loss = 0.001236 (epoch 349, PR #2828, W&B run libpwryz)
- **DrivAerML**: val/loss = 4.619% (epoch 256, PR #2691, W&B run k8qtsxxz)

Reproduce commands:
```bash
# TandemFoil baseline
cd target/icml2026 && python train.py --dataset tandemfoil --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999

# AirfRANS baseline
cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999

# DrivAerML baseline
cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT robin:

## Status Update

Found and fixed a bug during initial experiments: `torch.compile` wraps models as `OptimizedModule` which prefixes parameter names with `_orig_mod.`. The EMA shadow dict used original names, so `ema.update(compiled_model)` silently skipped all parameters — the shadow never updated, causing frozen val metrics during evaluation.

**Fix:** Added `_clean()` method to strip the `_orig_mod.` prefix in all EMA methods (update, store, copy_to, restore). Verified fix works — val metrics now properly change between epochs.

**Note:** This bug also exists in the original `EMA` class in `core/optim.py` (same name-mismatch issue), though it was never triggered because all baseline runs use `--no-use-ema`. Worth fixing upstream if EMA is being re-enabled.

All 6 experiments relaunched on GPUs 0-5. Will report full results when training completes.

**Comment by tcapelle** [comment]
STUDENT robin:

## Results

### Summary

EMA warmup (timm-style adaptive decay) is a **clear win on 2 of 3 datasets** (AirfRANS and TandemFoil), with decay=0.999 consistently outperforming decay=0.9999. DrivAerML is a negative result — the aggressive T_max=30 cosine cycling creates oscillations too wide for EMA to be useful.

**Bug fix included:** The original EMA class was silently broken with `torch.compile` — parameter names get prefixed with `_orig_mod.` by `OptimizedModule`, causing `ema.update()` to skip all parameters. Added `_clean()` method to strip the prefix. This bug also exists in `core/optim.py`'s `EMA` class (never triggered because baselines use `--no-use-ema`).

### AirfRANS — decay=0.999: NEW BEST (-41.2%)

| Metric | Baseline (no-EMA) | EMA 0.999 | EMA 0.9999 |
|---|---|---|---|
| **val_primary/surface_mse** | **0.001236** (ep349) | **0.000727** (ep206) | **0.001123** (ep275) |
| vs baseline | — | **-41.2%** | **-9.1%** |
| vs external target (0.0043) | -71.3% | **-83.1%** | -73.9% |

- W&B runs: [i1sevgt2](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/i1sevgt2) (0.999), [bz00wego](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/bz00wego) (0.9999)
- Config: 2L/256d/4H, AdamW lr=7e-4, T_max=5, gc=1.0, WD=1e-2, Fourier
- Commands:
  ```bash
  # decay=0.999 (WINNER)
  cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --ema-decay 0.999 --agent robin --wandb-name "robin/ema-warmup-airfrans-0.999" --wandb-group ema-warmup
  # decay=0.9999
  cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --agent robin --wandb-name "robin/ema-warmup-airfrans-0.9999" --wandb-group ema-warmup
  ```
- Peak memory: ~39 GiB (GPU 2), ~21 GiB (GPU 3)

### TandemFoil — decay=0.999: NEW BEST (-11.9%)

| Metric | Baseline (no-EMA) | EMA 0.999 | EMA 0.9999 |
|---|---|---|---|
| **val_primary/surface_pressure_mae** | **30.10** (ep157) | **26.52** (ep116) | **27.17** (ep105) |
| vs baseline | — | **-11.9%** | **-9.7%** |

- W&B runs: [nrn0q3ct](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/nrn0q3ct) (0.999), [3xi6cgx1](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/3xi6cgx1) (0.9999)
- Config: 3L/192d/3H, Lion lr=1.25e-4, T_max=10, gc=1.0, WD=1e-2, Fourier+physics, slices=64
- Commands:
  ```bash
  # decay=0.999 (WINNER)
  cd target/icml2026 && python train.py --dataset tandemfoil --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --ema-decay 0.999 --agent robin --wandb-name "robin/ema-warmup-tandem-0.999" --wandb-group ema-warmup
  # decay=0.9999
  cd target/icml2026 && python train.py --dataset tandemfoil --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --agent robin --wandb-name "robin/ema-warmup-tandem-0.9999" --wandb-group ema-warmup
  ```
- Peak memory: ~46 GiB (GPU 5), ~23 GiB (GPU 4)

### DrivAerML — NEGATIVE RESULT

| Metric | Baseline (no-EMA) | EMA 0.9999 | EMA 0.999 |
|---|---|---|---|
| **val_primary/surface_rel_l2_pct** | **4.619%** (ep256) | **9.749%** (ep60) | **19.41%** (ep22) |
| vs baseline | — | +111% | +320% |

- W&B runs: [3vgfvj7b](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/3vgfvj7b) (0.9999), [7fm3892g](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/7fm3892g) (0.999)
- Both runs plateaued early (ep60 and ep22 respectively) and never improved despite 200+ additional epochs
- Config: 4L/512d/8H, AdamW lr=5e-4, T_max=30, Fourier
- **Why it failed:** T_max=30 creates extreme oscillations (val swings 10%→67% between cosine peaks/troughs). The EMA accumulates weights from both good (trough) and bad (peak) phases, averaging them into mediocrity. The non-EMA model can instantaneously benefit from the deep troughs at cosine minima, but the EMA model lags behind and gets "pulled up" by peak-phase weights.
- Peak memory: ~6 GiB per GPU

### What happened

The EMA warmup hypothesis is **validated on datasets with moderate LR oscillations** (T_max=5 for AirfRANS, T_max=10 for TandemFoil). The adaptive decay formula `min(target, (1+step)/(10+step))` prevents early-training noise from poisoning the EMA, and once warmed up, the EMA acts as a de facto ensemble of late-training checkpoints.

**Key finding: decay=0.999 > decay=0.9999.** The lower target decay makes the EMA more responsive to recent weight updates, which is critical for tracking the improving model during long training runs. decay=0.9999 makes the EMA too "sticky" — it can't forget old weights fast enough.

**DrivAerML failure mechanism:** When cosine cycling is aggressive (T_max=30), the model weight trajectory oscillates wildly between good and bad states. The EMA, being a running average, cannot follow the sharp transitions to deep troughs. The non-EMA model benefits from being evaluated exactly at trough moments; the EMA model is always a blurred average.

### Suggested follow-ups

1. **Merge decay=0.999 as new default for AirfRANS and TandemFoil** — clear wins on both
2. **Fix the `_orig_mod.` bug in `core/optim.py`** upstream so the original EMA class also works with `torch.compile`
3. **Test EMA warmup on DrivAerML with smaller T_max** (e.g., T_max=10 or T_max=15) — the failure is likely T_max-specific, not dataset-specific
4. **Try decay=0.99** — given that 0.999 beat 0.9999, an even more responsive EMA might help further
5. **Compound with other recent wins** (e.g., T_max=10 for AirfRANS at 2L/256d) — EMA warmup is orthogonal to architecture/LR changes

*Note: All 6 experiments are still running (currently ~3h into 6h budget). Numbers above are best-so-far; final values may be even better. Check W&B runs for live tracking.*

**Comment by tcapelle** [comment]
ADVISOR: Merging — corrected EMA warmup is a major breakthrough on two datasets:

- **AirfRANS:** 0.000727 (W&B run i1sevgt2, decay=0.999, ep206) — **-41.2% vs baseline 0.001236**
- **TandemFoil:** 26.134 (W&B run nrn0q3ct, decay=0.999, ep123) — **-13.2% vs baseline 30.10**

The timm-style warmup formula eliminates the early-step EMA corruption that forced us to use --no-use-ema. Clean implementation with proper store/restore, _orig_mod prefix handling for compile compat, and ema_decay_actual logging.

DrivAerML did not benefit from EMA alone (9.749% at ep60 vs 4.619% baseline) but TF+AF wins are definitive. The shared recipe now includes corrected EMA (--ema-decay 0.999) for TF and AF.

**Comment by tcapelle** [comment]
STUDENT robin:

## Final Results (all 6 experiments completed, 360-min budget exhausted)

### Definitive final numbers

| Experiment | Best Val | At Epoch | Total Epochs | vs Baseline | W&B |
|---|---|---|---|---|---|
| **AirfRANS 0.999** | **0.000727** | ep206 | 734 | **-41.2%** | i1sevgt2 |
| **AirfRANS 0.9999** | **0.000869** | ep421 | 738 | **-29.7%** | bz00wego |
| **TandemFoil 0.999** | **22.96** | ep314 | 322 | **-23.7%** | nrn0q3ct |
| **TandemFoil 0.9999** | **23.87** | ep306 | 306 | **-20.7%** | 3xi6cgx1 |
| DrivAerML 0.9999 | 9.749% | ep60 | 455 | +111% | 3vgfvj7b |
| DrivAerML 0.999 | 19.41% | ep22 | 451 (diverged ep230) | +320% | 7fm3892g |

### Notes on divergence
- **AirfRANS 0.999** diverged at ep362 (same T_max=5 instability pattern as prior baselines). Best trough (0.000727 @ep206) is the definitive result.
- **DrivAerML 0.999** diverged to NaN at ep230.
- All other experiments ran to timeout without divergence.

### Key takeaway
EMA warmup with decay=0.999 delivers **23-41% improvements** on AirfRANS and TandemFoil. Both decay values beat baselines on both datasets. DrivAerML is a negative result due to T_max=30 oscillation severity. The `_orig_mod.` torch.compile bug fix is essential for EMA to work at all.
